src/log-store/src/kafka/worker.rs

```diff
@@ -0,0 +1,208 @@
+            ProduceResultHandle { receiver: rx },
+        )
+    }
+}
+
+pub(crate) struct DumpIndexRequest {
+    encoder: Arc<dyn IndexEncoder>,
+    sender: oneshot::Sender<()>,
+}
+
+impl DumpIndexRequest {
+    pub fn new(encoder: Arc<dyn IndexEncoder>) -> (DumpIndexRequest, oneshot::Receiver<()>) {
+        let (tx, rx) = oneshot::channel();
+        (
+            DumpIndexRequest {
+                encoder,
+                sender: tx,
+            },
+            rx,
+        )
+    }
+}
+
+pub(crate) struct TruncateIndexRequest {
+    region_id: RegionId,
+    entry_id: EntryId,
+}
+
+pub(crate) struct ProduceRequest {
+    region_id: RegionId,
+    batch: Vec<Record>,
+    sender: oneshot::Sender<ProduceResultReceiver>,
+}
+
+/// Receives the committed offsets when data has been committed to Kafka
+/// or an unrecoverable error has been encountered.
+pub(crate) struct ProduceResultHandle {
+    receiver: oneshot::Receiver<ProduceResultReceiver>,
+}
+
+impl ProduceResultHandle {
+    /// Waits for the data has been committed to Kafka.
+    /// Returns the **max** committed offsets.
+    pub(crate) async fn wait(self) -> Result<u64> {
+        self.receiver
+            .await
+            .context(error::WaitProduceResultReceiverSnafu)?
+            .wait()
+            .await
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct ProduceResultReceiver {
+    receivers: Vec<oneshot::Receiver<Result<Vec<i64>>>>,
+}
+
+impl ProduceResultReceiver {
+    fn add_receiver(&mut self, receiver: oneshot::Receiver<Result<Vec<i64>>>) {
+        self.receivers.push(receiver)
+    }
+
+    async fn wait(self) -> Result<u64> {
+        Ok(try_join_all(self.receivers)
+            .await
+            .into_iter()
+            .flatten()
+            .collect::<Result<Vec<_>>>()?
+            .into_iter()
+            .flatten()
+            .max()
+            .context(NoMaxValueSnafu)? as u64)
+    }
+}
+
+pub(crate) struct PendingRequest {
+    batch: Vec<Record>,
+    region_ids: Vec<RegionId>,
+    size: usize,
+    sender: oneshot::Sender<Result<Vec<i64>>>,
+}
+
+pub(crate) struct BackgroundProducerWorker {
+    pub(crate) provider: Arc<KafkaProvider>,
+    /// The [`ProducerClient`].
+    pub(crate) client: Arc<dyn ProducerClient>,
+    // The compression configuration.
+    pub(crate) compression: Compression,
+    // The running flag.
+    pub(crate) running: Arc<AtomicBool>,
+    /// Receiver of [ProduceRequest].
+    pub(crate) receiver: Receiver<WorkerRequest>,
+    /// Max batch size for a worker to handle requests.
+    pub(crate) request_batch_size: usize,
+    /// Max bytes size for a single flush.
+    pub(crate) max_batch_bytes: usize,
+    /// Collecting ids of WAL entries.
+    pub(crate) index_collector: Box<dyn IndexCollector>,
+}
+
+impl BackgroundProducerWorker {
+    pub(crate) async fn run(&mut self) {
+        let mut buffer = Vec::with_capacity(self.request_batch_size);
+        while self.running.load(Ordering::Relaxed) {
```

Comment:
There's no need to use a separate `AtomicBool` to control this background task. Simply make `OrderedBatchProducer::sender` an `Option<Sender<...>>` will be enough. Set `OrderedBatchProducer::sender` to None will enforce the recver here get a None which can not only wake up the await here but also break the loop/